Nil pointer: Network Idle Context Done

This one is coming from `doneCh := f.networkIdleCtx.Done()`.

I used Tom's script in #49 with `headless` mode `true`.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1c372dc]
default   [--------------------------------------] 1 VUs  00m05.5s/10m0s  0/1 iters, 1 per VU
goroutine 2536 [running]:
github.com/grafana/xk6-browser/common.(*Frame).startNetworkIdleTimer.func1()
	github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/common/frame.go:290 +0x5c
created by github.com/grafana/xk6-browser/common.(*Frame).startNetworkIdleTimer
	github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/common/frame.go:288 +0x135
2021/11/12 13:52:46 [ERROR] exit status 2
```

Running a similar issue. I have just installed K6 and XK6-Browser. When running a simple script with xk6-browser run command. I get below error.

Script:

```javascript
import launcher from "k6/x/browser";

const browser = launcher.launch('chromium', { headless: false });
const context = browser.newContext();
const page = context.newPage();

page.goto('https://www.google.com');
page.close();

browser.close();
```

Error:
```bash
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x238 pc=0x10196d9]

goroutine 1 [running]:
github.com/dop251/goja.(*Runtime).RunProgram.func1()
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/runtime.go:1262 +0x74
panic({0x116b0c0, 0x2124480})
	runtime/panic.go:1038 +0x215
github.com/dop251/goja.(*vm).try.func1()
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:512 +0x8eb
panic({0x116b0c0, 0x2124480})
	runtime/panic.go:1038 +0x215
github.com/grafana/xk6-browser/common.NewBrowser({0x15a0f08, 0xc0018e1da0}, 0xc0025b7720, 0xc000d00eb0, 0xc000ce3bc0)
	github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/common/browser.go:106 +0x159
github.com/grafana/xk6-browser/chromium.(*BrowserType).Launch(0xc0018e1d10, {0x15c00d8, 0xc0018e1b30})
	github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/chromium/browser_type.go:132 +0xf4b
github.com/grafana/xk6-browser.(*JSModule).Launch(0x0, {0xc000791260, 0x45d967}, {0x15c00d8, 0xc0018e1b30})
	github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/main.go:97 +0x6e
reflect.Value.call({0x113c480, 0xc000e2c180, 0x7fb836d99108}, {0x12fe4ba, 0x4}, {0xc0018e1b90, 0x2, 0xc00055feb0})
	reflect/value.go:543 +0x814
reflect.Value.Call({0x113c480, 0xc000e2c180, 0xc0018e1b30}, {0xc0018e1b90, 0x2, 0x2})
	reflect/value.go:339 +0xc5
github.com/dop251/goja.(*Runtime).wrapReflectFunc.func1({{0x15c00d8, 0xc0018e1530}, {0xc00138b120, 0x2, 0x6}})
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/runtime.go:1757 +0x259
github.com/dop251/goja.(*vm)._nativeCall(0xc002274600, 0xc0023aa1e0, 0x2)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:2658 +0x223
github.com/dop251/goja.call.exec(0x8, 0xc002274600)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:2630 +0x81b
github.com/dop251/goja.(*vm).run(0xc002274600)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:402 +0x8b
github.com/dop251/goja.(*vm).try(0xc0005603a8, 0x9eb005)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:518 +0x1a4
github.com/dop251/goja.(*vm).runTry(...)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/vm.go:523
github.com/dop251/goja.(*Runtime).RunProgram(0xc0025a4000, 0xc0018e0f00)
	github.com/dop251/goja@v0.0.0-20211022113120-dc8c55024d06/runtime.go:1277 +0x19e
go.k6.io/k6/js.(*Bundle).instantiate(0xc0002d02c0, {0x15c57e0, 0xc000710380}, 0xc0025a4000, 0xc000d00d70, 0x29)
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/js/bundle.go:327 +0x785
go.k6.io/k6/js.NewBundle({0x15c57e0, 0xc000710380}, 0xc00070e5e0, 0xc00072c000, {{{0x1, 0x0}}, {{{0x1303ce8, 0x8}, 0x0}}, 0xc00072c300, ...}, ...)
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/js/bundle.go:104 +0x45d
go.k6.io/k6/js.New(0x240, 0x0, 0xc0000bfb00, {{{0x1, 0x0}}, {{{0x1303ce8, 0x8}, 0x0}}, 0xc00072c300, {{0x0, ...}}, ...}, ...)
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/js/runner.go:83 +0x96
go.k6.io/k6/cmd.newRunner(0xc0001a5bc0, 0xc0001a5800, {0x12fdd70, 0xc0001a5500}, 0xc00073c0c0, {{{0x1, 0x0}}, {{{0x1303ce8, 0x8}, 0x0}}, ...}, ...)
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/cmd/run.go:398 +0xb4
go.k6.io/k6/cmd.newRunner(0x140, 0xc00070e5e0, {0x0, 0x0}, 0x270, {{{0x1, 0x0}}, {{{0x1303ce8, 0x8}, 0x0}}, ...}, ...)
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/cmd/run.go:396 +0x38d
go.k6.io/k6/cmd.getRunCmd.func1(0xc00071c780, {0xc000489710, 0x1, 0x1})
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/cmd/run.go:121 +0x556
github.com/spf13/cobra.(*Command).execute(0xc00071c780, {0xc0004896f0, 0x1, 0x1})
	github.com/spf13/cobra@v1.1.3/command.go:852 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc000702a00)
	github.com/spf13/cobra@v1.1.3/command.go:960 +0x3ad
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.1.3/command.go:897
go.k6.io/k6/cmd.Execute()
	go.k6.io/k6@v0.34.2-0.20211105090803-dae70006ca77/cmd/root.go:182 +0x47a
main.main()
	k6/main.go:13 +0x17
```

--- a/common/frame.go
+++ b/common/frame.go
@@ -126,9 +126,7 @@ type Frame struct {
 
 	loadingStartedTime time.Time
 
-	networkIdleMu       sync.Mutex
-	networkIdleCtx      context.Context
-	networkIdleCancelFn context.CancelFunc
+	networkIdleCh chan struct{}
 
 	inflightRequestsMu sync.RWMutex
 	inflightRequests   map[network.RequestID]bool
@@ -153,6 +151,7 @@ func NewFrame(ctx context.Context, m *FrameManager, parentFrame *Frame, frameID 
 		utilityExecutionContextCh: make(chan bool, 1),
 		inflightRequests:          make(map[network.RequestID]bool),
 		currentDocument:           &DocumentInfo{},
+		networkIdleCh:             make(chan struct{}),
 	}
 }
 
@@ -266,12 +265,9 @@ func (f *Frame) recalculateLifecycle() {
 }
 
 func (f *Frame) stopNetworkIdleTimer() {
-	f.networkIdleMu.Lock()
-	defer f.networkIdleMu.Unlock()
-	if f.networkIdleCancelFn != nil {
-		f.networkIdleCancelFn()
-		f.networkIdleCtx = nil
-		f.networkIdleCancelFn = nil
+	select {
+	case f.networkIdleCh <- struct{}{}:
+	default:
 	}
 }
 
@@ -279,16 +275,11 @@ func (f *Frame) startNetworkIdleTimer() {
 	if f.hasLifecycleEventFired(LifecycleEventNetworkIdle) || f.detached {
 		return
 	}
-	if f.networkIdleCancelFn != nil {
-		f.networkIdleCancelFn()
-	}
-	networkIdleCtx, networkIdleCancelFn := context.WithCancel(f.ctx)
-	f.networkIdleMu.Lock()
-	f.networkIdleCtx, f.networkIdleCancelFn = networkIdleCtx, networkIdleCancelFn
-	f.networkIdleMu.Unlock()
+	f.stopNetworkIdleTimer()
 	go func() {
 		select {
-		case <-networkIdleCtx.Done():
+		case <-f.ctx.Done():
+		case <-f.networkIdleCh:
 		case <-time.After(LifeCycleNetworkIdleTimeout):
 			f.manager.frameLifecycleEvent(f.id, LifecycleEventNetworkIdle)
 		}

--- a/common/frame.go
+++ b/common/frame.go
@@ -282,15 +282,13 @@ func (f *Frame) startNetworkIdleTimer() {
 	if f.networkIdleCancelFn != nil {
 		f.networkIdleCancelFn()
 	}
+	networkIdleCtx, networkIdleCancelFn := context.WithCancel(f.ctx)
 	f.networkIdleMu.Lock()
-	f.networkIdleCtx, f.networkIdleCancelFn = context.WithCancel(f.ctx)
+	f.networkIdleCtx, f.networkIdleCancelFn = networkIdleCtx, networkIdleCancelFn
 	f.networkIdleMu.Unlock()
 	go func() {
-		f.networkIdleMu.Lock()
-		doneCh := f.networkIdleCtx.Done()
-		f.networkIdleMu.Unlock()
 		select {
-		case <-doneCh:
+		case <-networkIdleCtx.Done():
 		case <-time.After(LifeCycleNetworkIdleTimeout):
 			f.manager.frameLifecycleEvent(f.id, LifecycleEventNetworkIdle)
 		}

